subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/GroovyRecompilationSpecProvider.java

```diff
@@ -98,47 +92,51 @@ public void initializeCompilation(JavaCompileSpec spec, RecompilationSpec recomp
 
     private void prepareFilePatterns(Set<File> filesToCompile, PatternSet classesToDelete, PatternSet filesToRecompilePatterns) {
         for (File file : filesToCompile) {
-            filesToRecompilePatterns.include(relativize(file));
-
-            Collection<String> classes = sourceFileClassNameConverter.getClassNames(file);
-            for (String staleClass : classes) {
-                String path = staleClass.replaceAll("\\.", "/");
-                classesToDelete.include(path.concat(".class"));
-            }
+            sourceDirs.relativize(file).ifPresent(relativeFile -> {
```

Comment:
We don't need to relativize any more since all the added files use already a relative path -> we should use `String` for the `filesToCompile` and rename the field/parameter to something like `relativeSourcePathsToCompile`.

Reply:
Done.